### PR TITLE
Added hvd.is_initialized() method to torch extension

### LINKED ISSUE #2011: Consistent way to check if horovod is shutdown ?
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.4
3. Horovod version: '0.19.3'
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**

Is there a consistent way to check if horovod is shutdown ?
For example,
```python
import horovod.torch as hvd
# hvd.rank() => raises ValueError: Horovod has not been initialized; use hvd.init().
hvd.init()
hvd.rank()  # for example gives 0
hvd.shutdown()
hvd.rank()  # still gives 0, however horovod is shutdown and I would expect the same behaviour as before init()
```

### PR DESCRIPTION
**Pull request by vfdev-5**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes #2011 

- Added hvd.is_initialized() method to torch extension

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by vfdev-5** [comment]
@tgaddair I added `is_initialized()` to basics and to torch extension. If this is something like that you think about I can add that to other extensions.

Function `is_initialized` return the state of `horovod_global.initialization_done` instead of `initialize_flag` which seems like we can not get the state without modifying it...

**Comment by vfdev-5** [comment]
@tgaddair could you, please, review the code and comment out if failing tests are related to the modifications ?

**Comment by tgaddair** [comment]
Hey @vfdev-5, thanks for the PR!  Looks like one the tests (test_horovod_join_allgather) is consistently failing, so I do think it is related to this PR:

```
*** The MPI_Query_thread() function was called after MPI_FINALIZE was invoked.
--
  | *** This is disallowed by the MPI standard.
  | *** Your MPI job will now abort.
  | [51389e92e649:2843] [0] func:/usr/local/lib/libopen-pal.so.40(opal_backtrace_buffer+0x20) [0x7fc51b97b090]
  | [51389e92e649:2843] [1] func:/usr/local/lib/libmpi.so.40(ompi_mpi_abort+0x258) [0x7fc51e34c4f8]
  | [51389e92e649:2843] [2] func:/usr/local/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xfd) [0x7fc51e33cf0d]
  | [51389e92e649:2843] [3] func:/usr/local/lib/libmpi.so.40(MPI_Query_thread+0x52) [0x7fc51e37b6f2]
  | [51389e92e649:2843] [4] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common10MPIContext10InitializeERKSt6vectorIiSaIiEERNS0_17MPIContextManagerE+0x1ed) [0x7fc4bd3f51ad]
  | [51389e92e649:2843] [5] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x4de86) [0x7fc4bd3a7e86]
  | [51389e92e649:2843] [6] func:/usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so(+0x3811f) [0x7fc525ca011f]
  | [51389e92e649:2843] [7] func:/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7fc52f5626db]
  | [51389e92e649:2843] [8] func:/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7fc52f89b88f]
  | [51389e92e649:2843] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
  | *** The MPI_Query_thread() function was called after MPI_FINALIZE was invoked.
  | *** This is disallowed by the MPI standard.
  | *** Your MPI job will now abort.
  | [51389e92e649:2844] [0] func:/usr/local/lib/libopen-pal.so.40(opal_backtrace_buffer+0x20) [0x7fe0959f8090]
  | [51389e92e649:2844] [1] func:/usr/local/lib/libmpi.so.40(ompi_mpi_abort+0x258) [0x7fe0961ab4f8]
  | [51389e92e649:2844] [2] func:/usr/local/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xfd) [0x7fe09619bf0d]
  | [51389e92e649:2844] [3] func:/usr/local/lib/libmpi.so.40(MPI_Query_thread+0x52) [0x7fe0961da6f2]
  | [51389e92e649:2844] [4] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common10MPIContext10InitializeERKSt6vectorIiSaIiEERNS0_17MPIContextManagerE+0x1ed) [0x7fe0964f91ad]
  | [51389e92e649:2844] [5] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x4de86) [0x7fe0964abe86]
  | [51389e92e649:2844] [6] func:/usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so(+0x3811f) [0x7fe09b59b11f]
  | [51389e92e649:2844] [7] func:/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7fe0a3b096db]
  | [51389e92e649:2844] [8] func:/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7fe0a3e4288f]
  | [51389e92e649:2844] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

File: test/test_torch.py [hunk 1]
```diff
@@ -85,14 +85,20 @@ def test_horovod_reinit(self):
...
         rank, size = hvd.rank(), hvd.size()
-
         hvd.shutdown()
         hvd.init()
         rank2, size2 = hvd.rank(), hvd.size()
 
         assert rank == rank2
         assert size == size2
 
+    def test_horovod_is_initialized(self):
```
Comment (tgaddair) [inline test/test_torch.py:95]: I suspect that the order of the tests is causing the failure due to state carryover between tests.  Are you able to reproduce this locally?  What if you moved this test to the bottom of this file?

Comment (vfdev-5) [reply test/test_torch.py:95]: Yes, you are right, the problem is that the test left hvd in a shutdown state. And similarly to `test_horovod_reinit` we need to shutdown and reinit only if gloo is used. I'll send a fix now.

File: test/test_torch.py [hunk 2]
```diff
@@ -85,14 +85,28 @@ def test_horovod_reinit(self):
...
+        """Test that is_initialized returned by hvd.is_initialized() is correct."""
+        hvd.init()
+        assert hvd.is_initialized()
+
+        gloo_rank = int(os.getenv('HOROVOD_RANK', -1))
+        is_mpi = gloo_rank == -1
+        if is_mpi:
+            # Only applies for Gloo
```
Comment (tgaddair) [inline test/test_torch.py:103]: Can you change this comment to make it more clear:

```
# Horovod cannot be re-initialized after shutdown when using MPI, so this test can only be done using the Gloo controller
```

Comment (vfdev-5) [reply test/test_torch.py:103]: Sure !

**Comment by vfdev-5** [comment]
@tgaddair  tests are passing now 🎉

**Pull request merged**